File: 10-heap/00215-Kth_Largest_Element_In_An_Array/00215-kth_largest_element_in_an_array.cpp
// 215: Kth Largest Element In An Array
// https://leetcode.com/problems/kth-largest-element-in-an-array


#include <iostream>
#include <queue>
#include <vector>

using namespace std;


class Solution {
public:
    // SOLUTION
    int findKthLargest(vector<int>& nums, int k) {
        priority_queue<int, vector<int>, greater<int>> pq;
        
        for (int num : nums) {
            pq.push(num);
            if (pq.size() > k) {
                pq.pop();
            }
        }
        
        return pq.top();
    }
};


int main() {
    Solution o;

    // INPUT
    vector<int> nums = {3,2,1,5,6,4};
    int k = 2;

    // OUTPUT
    auto result = o.findKthLargest(nums, k);
    cout << result << endl;

    return 0;
}

File: 10-heap/00215-Kth_Largest_Element_In_An_Array/00215-kth_largest_element_in_an_array.java
// 215: Kth Largest Element In An Array
// https://leetcode.com/problems/kth-largest-element-in-an-array


import java.util.PriorityQueue;

class Solution {
    // SOLUTION
    public int findKthLargest(int[] nums, int k) {
        PriorityQueue<Integer> pq = new PriorityQueue<Integer>(k + 1);

        for(int num : nums) {
            pq.add(num);
            if (pq.size() > k) {
                pq.poll();
            }
        }

        return pq.poll();
    }

    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        int[] nums = {3,2,1,5,6,4};
        int k = 2;

        // OUTPUT
        var result = o.findKthLargest(nums, k);
        System.out.println(result);
    }
}